### ISSUE TITLE
Extracting secondary structure (DSSP)

### ISSUE THREAD

**Issue by Bedinka** [Bedinka/PPI-prediction-with-Deep-Learning-Language-Model-#18]
# PPI Analysis with Deep Learning Language Model Project

## Description

<p style="box-sizing: border-box; line-height: 24px; margin: 0px 0px 24px; font-size: 16px; color: rgb(64, 64, 64); font-family: Lato, proxima-nova, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(252, 252, 252); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">DSSP (<strong style="box-sizing: border-box; font-weight: bold;">D</strong>efine<span> </span><strong style="box-sizing: border-box; font-weight: bold;">S</strong>econdary<span> </span><strong style="box-sizing: border-box; font-weight: bold;">S</strong>tructure of<span> </span><strong style="box-sizing: border-box; font-weight: bold;">P</strong>roteins) is the standard method used to assign secondary structure annotations to a protein structure. DSSP utilizes the atomic coordinates of a structure to assign the secondary codes, which are:</p>

![Screenshot from 2024-03-12 11-58-09](https://github.com/lionking0000/YangLabIntern/assets/105316730/fa73faf0-255e-49af-8b59-d789fc510e67)

<p style="box-sizing: border-box; line-height: 24px; margin: 0px 0px 24px; font-size: 16px; color: rgb(64, 64, 64); font-family: Lato, proxima-nova, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(252, 252, 252); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Furthermore, DSSP calculates geometric properties such as the phi and psi angles between residues and solvent accessibilities.</p>

---

### Problem Statement

---

### Proposed Solution

Proposed imported module:
- PyDssp:  https://pypi.org/project/pydssp/

---

### Expected Behavior (for bugs)

Create a Hbond matrixes, secondary structure matrix, secondary structure as indexes matrix, one hot representation passit back to the corresponding chain object.
---

### Actual Behavior (for bugs)
[Explain what actually happened when following the steps above.]

---

### Additional Context
[Add any additional context, screenshots, or information that may be relevant to the issue.]

---

### Related Code
'
def dssp(chain_CA, pdb):
    for chain in chain_CA.values():
          ## atoms should be 4 (N, CA, C, O) or 5 (N, CA, C, O, H)
              coord = torch.tensor(pydssp.read_pdbtext(open(pdb, 'r').read()))
          # hydrogene bond matrix
              hbond_matrix = pydssp.get_hbond_map(coord)
              #print(hbond_matrix.shape) # should be (batch, length, length)
          # getting scondary struct 
              dssp_struct = pydssp.assign(coord, out_type='c3')
              chain.dssp_struct = dssp_struct
          ## output is batched np.ndarray of C3 annotation, like ['-', 'H', 'H', ..., 'E', '-']
          # To get secondary str. as index
              dssp_index = pydssp.assign(coord, out_type='index')
              chain.dssp_indes = dssp_index
          ## 0: loop,  1: alpha-helix,  2: beta-strand
          # To get secondary str. as onehot representation
              dssp_onhot = pydssp.assign(coord, out_type='onehot')
              chain.dssp_onehot = dssp_onhot
          ## dim-0: loop,  dim-1: alpha-helix,  dim-2: beta-strand
              #print(len(dssp_struct))
              #print(hbond_matrix)
              output_file = "./" + str(chain.prot_id) + ".dssp.txt"
              with open(output_file, 'w') as f:
                  f.write(f"{hbond_matrix}\n{dssp_struct}\n{dssp_index}\n{dssp_onhot}\n")
'

---

### Priority
[Set the priority level for this issue (e.g., high, medium, low).]

---

### Due Date
[Specify a due date, if necessary.]

---
### Checklist for coding 
- [x] Modul configuration.
- [x] Seperate the interacting protein chains from each file. 
- [x] Calculation of each DSSP in each chain seperatelly.
- [x] Calculating Hbond matrix of each chain.

### Checklist for tasks 
- [x] Reviewed the problem statement and proposed solution.
- [x] Environmental dependecy solutions.
- [x] Provided expected and actual behavior matches.
- [x] Included all relevant context and code snippets.
- [x] Commited.
